internal/providers/aws/aws.go

```diff
@@ -73,3 +75,49 @@ func Init(f *resource.Fetcher) error {
 	f.S3RegionHint = regionHint
 	return nil
 }
+
+// FetchFromAWSMetadata fetches metadata from the `IMDSv2` service if its configured, else
+// it will fall back to `IMDSv1`. By default, the value of url is an IMDSv1 endpoint
+// i.e. http://169.254.169.254/2019-10-01/user-data.
+func FetchFromAWSMetadata(u url.URL, opts resource.FetchOptions, f *resource.Fetcher) ([]byte, error) {
+	userdataLatestURL := url.URL{
+		Scheme: "http",
+		Host:   "169.254.169.254",
+		Path:   "latest",
+	}
+	token, err := fetchAWSIMDSV2Token(userdataLatestURL, f)
+	if err == resource.ErrIMDSV2 {
+		// Do nothing
+		f.Logger.Info("IMDSv2 service is unavailable; falling back to IMDSv1")
+	} else if err != nil {
+		return nil, err
+	} else if len(token) > 0 && err == nil {
```

Comment:
`&& err == nil` is redundant since the previous case caught non-nil errors.

Do we need to check for zero-length tokens, or can we unconditionally assume we got a good token if there was no error?